### ISSUE TITLE
[FEATURE] Improve integration test harness

### ISSUE THREAD

**Issue by shonfeder** [informalsystems/apalache#192]
We currently just use a bash script for running our integration tests. This is brittle, requires reading and writing bash, and has a lot of limits (e.g., cannot be integrated into proper incremental build/test cycle, has no support for running specific tests). It would improve our CI and testing discipline to use a more mature test harness for our integration tests.

**Comment by konnov** [comment]
That is true. Any suggestions? There must be frameworks for testing CLI programs, but I don't know them. In any case, they should not be significantly harder than bash, e.g., we should not write tonnes of xml.

**Comment by shonfeder** [comment]
The smoothest integration tests for CLI I know of are via cram-style tests. E.g., using `mdx`, each logically related set of CLI integration tests could be in one markdown file and would look something like the following:

---

## Can check HourClock.tla specification

```sh
$ apalache-mc --length=10 --init=HCini --next=HCnxt --inv=TypeOK HourClock.tla
...
The outcome is: NoError
...
```

## Y2k Override fails with a length of 30

```sh
$ apalache-mc --length=10 --init=HCini --next=HCnxt --inv=TypeOK HourClock.tla
...
The outcome is: Error
...
[1]
```

---

The tests run the command following the `$`, and match the output with the following lines (`...` are wildcards, and `[1]` indicates return codes). If the tests fail, the do so with a diff presenting the actual vs. the expected output. If the actual output is what should be expected, the user executes `make promote` and the test output gets updated. This makes writing the tests very simple. You just write some shell commands in a markdown file, then promote the correct results (substituting in wildcards wherever you may have non-deterministic outcomes).

For a working example, you can see the integration tests for `themis-render`: https://gitlab.com/informalsystems/shared/research-and-development/plaintext/themis-render/-/blob/master/test/template-generation.md (that markdown file *is* the test suite).

A nice thing here is that we can run mdx on our documentation too, turning that into integration tests.

We could also use the python [cram](https://pypi.org/project/cram) package. It doesn't have the same support for markdown, and is implemented in a less nice language, but should still support nice, declarative integration tests.

**Comment by shonfeder** [comment]
if cram style tests aren't appealing, I'm also happy to budget some time to do a little survey of other available options!

**Comment by lemmy** [comment]
FWIW: TLC originally had a test suite that matched TLC's output. It was very brittle, which is why nobody ran the suite for years.  In the end, I converted the suite into something on top of JUnit, which still sucks but at least runs frequently and doesn't produce a ton of false failures.

**Comment by shonfeder** [comment]
Yeah, good point. Matching on output (which is what we're doing now in our bash file, btw) does bring some brittleness. That is mitigated to an extent with cram style tests by:

1. Easy use of wildcards that allow neutralizing any non-deterministic or irrelevant output.
2. Fully automated promotion of corrected output. So if something essential changes in the output, and the mach starts failing, you just review the diff to be sure the new output is what you want, and then execute one command to update.

I'm not sure if there's something particular about our output here that would make this worse. But my experience using this approach with [dune](https://github.com/ocaml/dune/tree/master/test/blackbox-tests/test-cases) has been pretty nice.

**Comment by shonfeder** [comment]
Thinking on @lemmy's point, a more principled approach to integration tests might be through a feature providing serialized output from apalache. If we add something like a `--json` flag, which causes apalache to output machine readible data, then it should be easy write more robust integration tests that check the behaviors we want. 

This could also be a help in the bench mark tests.

**Comment by konnov** [comment]
I like the approach with cram. I have a similar experience that we rarely look at these tests. But they are running with every Travis build, so we actually caught several non-obvious breaking changes with them. We should actually contribute to these tests more often.

**Comment by konnov** [comment]
Writing apalache output in a structured format with json looks like overkill to me. We can do a lot with regular expressions, assuming that the checked output is stable.

**Comment by shonfeder** [comment]
I'd argue that there's a lot of value to having machine readable output, but I won't do that here :D  I'm happy to go with cram style tests, and think it can add some value both to our testability and since it doubles as documentation :)

**Comment by konnov** [comment]
@shonfeder if you like to go ahead with that, please do. It will improve the infrastructure.

**Comment by konnov** [comment]
It looks like we will need integration tests for the type checker as well. It would be cool to go ahead with cram.

**Comment by shonfeder** [comment]
I'm going to situate this work item as dependent on #220, since we may otherwise need to do re-work to get this in place.
